File: src/algorithms/searching/binary_search/exercises/typescript/binary_search.ts
export const binarySearch = (nums: number[], target: number): number | null => {
  let left = 0
  let right = nums.length - 1
  while (left <= right) {
    const mid = Math.floor((left + right) / 2)

    if (nums[mid] === target) {
      return mid
    }

    if (target > nums[mid]) {
      left = mid + 1
    } else {
      right = mid - 1
    }
  }
  return null
}

/*
 BINARY SEARCH ARRAY
 *** Description
 Given a sorted array and a value, determine if the value is in the array using the binary search (divide and conquer) method.
 *** Exercises
 Write a function that takes a sorted array and a value and returns the index of the value in the array. Return null if the value is not found in the array. What is the time complexity?
 Extra credit: Implement the function both iteratively and recursively.
 */

File: src/algorithms/searching/binary_search/exercises/python/binary_search.py
# For a binary search to work a list must be sorted


def binary_search(arg: list, item: int):
    low = 0
    high = len(arg) - 1
    while low <= high:
        mid = low + high
        guess = arg[mid]
        if guess == item:
            return mid
        if guess > item:
            high = mid - 1
        else:
            low = mid + 1
    return None
